Source/JavaScriptCore/runtime/ArrayBufferView.cpp

```diff
@@ -41,10 +44,41 @@ ArrayBufferView::ArrayBufferView(
         m_baseAddress = BaseAddress(static_cast<char*>(m_buffer->data()) + m_byteOffset, byteLength);
 }
 
-ArrayBufferView::~ArrayBufferView()
+template<typename Visitor> constexpr decltype(auto) ArrayBufferView::visitDerived(Visitor&& visitor)
 {
-    if (!m_isDetachable)
-        m_buffer->unpin();
+    switch (m_type) {
+    case TypedArrayType::NotTypedArray:
+    case TypedArrayType::TypeDataView:
+        return std::invoke(std::forward<Visitor>(visitor), *static_cast<DataView*>(this));
+#define DECLARE_TYPED_ARRAY_TYPE(name) \
+    case TypedArrayType::Type ## name: \
+        return std::invoke(std::forward<Visitor>(visitor), *static_cast<name##Array*>(this));
+    FOR_EACH_TYPED_ARRAY_TYPE_EXCLUDING_DATA_VIEW(DECLARE_TYPED_ARRAY_TYPE)
+#undef DECLARE_TYPED_ARRAY_TYPE
+
+    }
+    ASSERT_NOT_REACHED();
+}
+
+template<typename Visitor> constexpr decltype(auto) ArrayBufferView::visitDerived(Visitor&& visitor) const
+{
+    return const_cast<ArrayBufferView&>(*this).visitDerived([&](auto& value) {
+        return std::invoke(std::forward<Visitor>(visitor), std::as_const(value));
+    });
+}
+
+JSArrayBufferView* ArrayBufferView::wrap(JSGlobalObject* lexicalGlobalObject, JSGlobalObject* globalObject)
+{
+    return visitDerived([&](auto& derived) { return derived.wrapImpl(lexicalGlobalObject, globalObject); });
+}
+
+void ArrayBufferView::operator delete(ArrayBufferView* value, std::destroying_delete_t)
+{
+    value->visitDerived([](auto& value) {
+        using T = std::decay_t<decltype(value)>;
```

Comment:
Seems to me that putting this all into the `freeAfterDestruction` line would be just fine without a named type: `std::decay_t<decltype(value)>::freeAfterDestruction(&value)`, unless maybe that doesn’t compile for some reason?

Reply:
I think this is better, here is the reason. Currently, this class is just a RefCounted. So we do not need to call a specific free function. But for example ISO_ALLOCATED ones need to call specific free functions, and this way can call appropriate free functions defined per class. And at the same time, if they are the same, since `freeAfterDestruction` is inline function, we expect that the generated code will be the same to call it outside of this `visitDerived`. :)
So, this way can prevent introducing a bug when ISO_ALLOCATED is used, while this way does not regress things for the other cases, so I think just using this style for all would be simpler rule.

Reply:
You may have misunderstood my feedback. I was suggesting a different format for the code where it is all on one line, not omitting the use of `std::decay_t<decltype(value)>` entirely. Further, I would encourage you to add the `std::decay_t<decltype(value)>` idiom to existing destroying delete implementations that should have it (maybe just the one).

Reply:
Ah! I see. Thanks, that makes sense